### PR TITLE
feat(storage): batch apply trivial-move task

### PR DESCRIPTION
**Pull request by Li0k**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Since, each etcd trx commit consumes more time. For the trivial-move task, each sst will use one etcd trx to complete, and frequent trivial-moves may block the compaction event_loop.

In this PR, batch submission of trivial tasks to etcd is introduced to alleviate the above problems. In order to achieve this feature, some refactorings have been carried out:

1. Generate a version delta based on multiple tasks
2. `apply_version_delta` supports multiple tasks included in a single version delta
3. Code simplification and refactoring

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/meta/src/hummock/manager/mod.rs [hunk 1]
```diff
@@ -777,23 +777,66 @@ impl HummockManager {
...
 
         let start_time = Instant::now();
-        // StoredIdGenerator already implements ids pre-allocation by ID_PREALLOCATE_INTERVAL.
-        let task_id = self
-            .env
-            .id_gen_manager()
-            .generate::<{ IdCategory::HummockCompactionTask }>()
-            .await?;
+
+        let trace_compact_task = |compact_task: &CompactTask, cost_time: Duration| {
```
Comment (Little-Wallace) [inline src/meta/src/hummock/manager/mod.rs:794]: could this be moved to another independent method?

File: src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs [hunk 2]
```diff
@@ -642,18 +636,59 @@ impl HummockLevelsExt for Levels {
...
+                break;
+            }
+        }
+    }
+
+    fn apply_compact_ssts(&mut self, group_deltas: &GroupDeltas) {
+        let mut has_delete_sst_levels = false;
+        for group_delta in group_deltas.get_group_deltas() {
+            if let DeltaType::IntraLevel(version_delta) = group_delta.get_delta_type().unwrap() {
+                let mut delete_sst_levels = Vec::with_capacity(1);
```
Comment (Little-Wallace) [inline src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs:672]: Could we collect all insert table and delete sstable ids for every group-delta? 
I means that we can place the variable `delete_sst_levels` outside this loop and change it to `HashMap<u32, HashSet<u64>>`

File: src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs [hunk 3]
```diff
@@ -33,41 +33,20 @@ use crate::prost_key_range::KeyRangeExt;
 use crate::{can_concat, CompactionGroupId, HummockSstableId, HummockSstableObjectId};
 
 pub struct GroupDeltasSummary {
```
Comment (Little-Wallace) [inline src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs:35]: We can store an array of `IntralLevelDelta` here.

File: src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs [hunk 4]
```diff
@@ -612,7 +604,9 @@ pub trait HummockLevelsExt {
     fn get_level(&self, idx: usize) -> &Level;
     fn get_level_mut(&mut self, idx: usize) -> &mut Level;
     fn count_ssts(&self) -> usize;
-    fn apply_compact_ssts(&mut self, summary: GroupDeltasSummary);
+    fn apply_compact_ssts(&mut self, group_deltas: &GroupDeltas);
```
Comment (Little-Wallace) [inline src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs:607]: ditto.
Here we can pass `Vec<IntraLevelDelta>` directly.

File: src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs (hunk 3 above)
Comment (Li0k) [reply src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs:37]: yes, fix it

File: src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs (hunk 2 above)
Comment (Li0k) [reply src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs]: It seems that only group_deltas of the same cg will be submitted each time, so I don't think group_deltas grouping is needed here anymore.

Can we keep it in the current pr and change it after breaking this assumption?

File: src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs [hunk 5]
```diff
@@ -532,39 +591,15 @@ impl HummockVersionUpdateExt for HummockVersion {
                 version_delta.max_committed_epoch,
                 self.max_committed_epoch
             );
+
             if self.max_committed_epoch < version_delta.max_committed_epoch {
                 // `max_committed_epoch` increases. It must be a `commit_epoch`
```
Comment (zwang28) [inline src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs:596]: Please help to fix the comment: It must be a `commit_epoch` or compaction group metadata change.

**Review by zwang28 — APPROVED** [review]
LGTM.

File: src/meta/src/hummock/manager/mod.rs [hunk 6]
```diff
@@ -850,60 +847,66 @@ impl HummockManager {
...
-                compact_task.target_level,
-                start_time.elapsed()
+
+            let compact_task = compact_status.get_compact_task(
+                current_version.get_compaction_group_levels(compaction_group_id),
+                task_id as HummockCompactionTaskId,
+                &group_config,
+                &mut stats,
+                selector,
+                table_id_to_option.clone(),
```
[the remark this answers is not in the document]
Comment (Little-Wallace) [reply src/meta/src/hummock/manager/mod.rs:866]: Agree

**Pull request closed without merging**